File: go/0380-insert-delete-getrandom-o1.go
package main

import "math/rand"

/*
https://leetcode.com/problems/insert-delete-getrandom-o1
Difficulty: Medium
TC: O(1)
SC: O(n)

Here we maintain a seperate list just because to use it in the GetRandom() function. Maps are unordered data structures, so we must convert it's keys to a list everytime we want to get a random value which is an O(n) operation. In order to do the same operation at O(1) constant time we must use a seperate list.
*/

type RandomizedSet struct {
	numsMap  map[int]int
	numsList []int
}

func Constructor_380() RandomizedSet {
	return RandomizedSet{make(map[int]int, 0), make([]int, 0)}
}

func (this *RandomizedSet) Insert(val int) bool {
	_, ok := this.numsMap[val]
	if ok {
		return false
	}
	this.numsList = append(this.numsList, val)
	this.numsMap[val] = len(this.numsList) - 1

	return true
}

func (this *RandomizedSet) Remove(val int) bool {
	index, ok := this.numsMap[val]
	if !ok {
		return false
	}

	lastIndex := len(this.numsList) - 1
	lastElement := this.numsList[lastIndex]

	// Move last element to the position of element being removed
	this.numsList[index] = lastElement
	this.numsList = this.numsList[:lastIndex]

	// Update map to reflect new position of moved element
	this.numsMap[lastElement] = index

	delete(this.numsMap, val)
	return true
}

func (this *RandomizedSet) GetRandom() int {
	return this.numsList[rand.Intn(len(this.numsList))]
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */

File: python/0380-insert-delete-getrandom-o1.py
"""
    https://leetcode.com/problems/insert-delete-getrandom-o1/
    Difficuilty: Medium

    Time complexity: O(1)
    Space complexity: O(n)
    Tags: list, hashmap

    Solution:
        - We keep a seperate list just because to use it in the GetRandom() function.
        - Maps are unordered data structures, so we must convert it's keys to a list everytime we want to get a random value which is an O(n) operation, it's also not guaranteed to have an even proablility using map. In order to do the same operation at O(1) constant time we must use a seperate list.
        - On removal of an element from list in O(1) time, we swap the index being removed with last item.
"""

import random


class RandomizedSet:

    def __init__(self):
        self.map = {}
        self.list = []

    def insert(self, val: int) -> bool:
        if val in self.map:
            return False
        self.list.append(val)
        self.map[val] = len(self.list) - 1
        return True
        

    def remove(self, val: int) -> bool:
        if val not in self.map:
            return False
        
        index = self.map[val]
        lastIndex = len(self.list) - 1
        lastElement = self.list[lastIndex]

        # swap last index with removed index 
        self.list[index] = lastElement
        self.list = self.list[:lastIndex]

        # Update map with the new position of the moved element
        self.map[lastElement] = index

        del self.map[val]
        return True

    def getRandom(self) -> int:
        return self.list[random.randint(0, len(self.list) - 1)]
        


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()